# DGEAnalysis_ShinyApps

These are R shiny applications for analyzing count data produced by genomic sequencing technologies, such as RNA-seq. The applications guide users through biological data assessment, processing and analysis. The different analyses available include differential expression (DE), network, and functional analysis.

## Download Instructions

The latest version of this application may be downloaded from [this repository](https://github.com/ElizabethBrooks/DGEAnalysis_ShinyApps/).

To download the code onto a local computer or server space, click the green code button near the top of the page and copy the link. Then, in the terminal:

git clone https://github.com/ElizabethBrooks/DGEAnalysis_ShinyApps.git

## Tutorials

### Web

Helpful information for downloading the applications in this repository can be found in the tutorial [GitHub Version Control Quick Start Guide](https://morphoscape.wordpress.com/2024/02/28/github-version-control-quick-start-guide/) 

A tutorial of the biostatistical analysis performed in this application is provided in [Downstream Bioinformatics Analysis of Omics Data with edgeR](https://morphoscape.wordpress.com/2022/08/09/downstream-bioinformatics-analysis-of-omics-data-with-edger/).

Gene tables were may be created from RNA-seq data as described in [Bioinformatics Analysis of Omics Data with the Shell & R](https://morphoscape.wordpress.com/2022/07/28/bioinformatics-analysis-of-omics-data-with-the-shell-r/).


## PDF 

The tutorials for using the applications or creating scripts for the different analyses can be found in the [tutorials](https://github.com/ElizabethBrooks/DGEAnalysis_ShinyApps/tree/main/tutorials) folder of this repository.

## Example Data Sets

Example gene counts and experimental design tables are also provided in the [data](https://github.com/ElizabethBrooks/DGEAnalysis_ShinyApps/tree/main/data) folder of this repository.

An example RNA-seq data set may be obtained from [ScienceDirect](https://www.sciencedirect.com/science/article/pii/S0147651319302684) and [NCBI](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA504739/).

## Acknowledgements

The applications were created by [Elizabeth Brooks](https://www.linkedin.com/in/elizabethmbrooks/).

This project was funded by the National Science Foundation grant "Collaborative Research: EDGE FGT: Genome-wide Knock-out mutant libraries for the microcrustacean Daphnia"  (2220695/2324639  to Sen Xu and 2220696 to Michael E. Pfrender).

## Analysis Pipeline 
![DGE Analysis Pipeline](RNASeqWorkflow_1Sep2023.jpg)

## References

